x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -71,34 +77,131 @@ export class EnterpriseSearchRequestHandler {
-          throw new Error('Cannot authenticate Enterprise Search user');
+          return this.handleAuthenticationError(response);
         }
 
+        // Handle 400-500+ responses from the Enterprise Search server
         const { status } = apiResponse;
-        const json = await apiResponse.json();
+        if (status >= 500) {
+          return this.handleServerError(response, apiResponse, url);
+        } else if (status >= 400) {
+          return this.handleClientError(response, apiResponse);
+        }
 
-        if (hasValidData(json)) {
-          return response.custom({ statusCode: status, body: json });
-        } else {
-          this.log.debug(`Invalid data received from <${url}>: ${JSON.stringify(json)}`);
-          throw new Error('Invalid data received');
+        // Check returned data
+        const json = await apiResponse.json();
+        if (!hasValidData(json)) {
+          return this.handleInvalidDataError(response, url, json);
         }
+
+        // Pass successful responses back to the front-end
+        return response.custom({ statusCode: status, body: json });
       } catch (e) {
-        const errorMessage = `Error connecting to Enterprise Search: ${e?.message || e.toString()}`;
+        // Catch connection/auth errors
+        return this.handleConnectionError(response, e);
+      }
+    };
+  }
 
-        this.log.error(errorMessage);
-        if (e instanceof Error) this.log.debug(e.stack as string);
+  /**
+   * Attempt to grab a usable error body from Enterprise Search - this isn't
+   * always possible because some of our internal endpoints send back blank
+   * bodies, and sometimes the server sends back Ruby on Rails error pages
+   */
+  async getErrorResponseBody(apiResponse: Response) {
+    const { statusText } = apiResponse;
+    const contentType = apiResponse.headers.get('content-type') || '';
 
-        return response.customError({ statusCode: 502, body: errorMessage });
-      }
+    // Default response
+    let body: IErrorResponse = {
+      message: statusText,
+      attributes: { errors: [statusText] },
     };
+
+    try {
+      if (contentType.includes('application/json')) {
+        // Try parsing body as JSON
+        const json = await apiResponse.json();
+
+        // Some of our internal endpoints return either an `error` or `errors` key,
+        // which can both return either a string or array of strings ¯\_(ツ)_/¯
+        const errors = json.error || json.errors || [statusText];
+        body = {
+          message: errors.toString(),
+          attributes: { errors: Array.isArray(errors) ? errors : [errors] },
+        };
+      } else {
+        // Try parsing body as text/html
+        const text = await apiResponse.text();
+        if (text) {
+          body = {
+            message: text,
+            attributes: { errors: [text] },
+          };
+        }
+      }
+    } catch {
+      // Fail silently
```

Comment:
Do we have a test for this? It'd be cool to document in a test what happens if we get a body that says it's json but we can't actually parse it.

Reply:
Great call, I'll add one!

Reply:
Addressed in https://github.com/elastic/kibana/pull/77258/commits/9bf80691dd228d336064eb1202fcc9eb9a1d86c5